### PR TITLE
fixes for mips cross compiling toolchains

### PR DESCRIPTION
**Pull request by Mic92**
Those don't evaluate.

###### Description of changes

Current status: still building. Testing here: https://github.com/nix-community/cross-toolchains.nix/pull/19

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Mic92** [comment]
Exception: error: Unknown CPU type: mipsisa32r6; type: nix::ThrownError
see https://hercules-ci.com/github/nix-community/cross-toolchains.nix/jobs/75

**Comment by Mic92** [comment]
cc @a-m-joseph

File: lib/systems/examples.nix [hunk 1]
```diff
@@ -98,20 +98,14 @@ rec {
   # can execute on 32bit chip
   mips-linux-gnu                = { config = "mips-linux-gnu";                } // platforms.gcc_mips32r2_o32;
```
Comment (Mic92) [inline lib/systems/examples.nix:99]: These also don't build ... https://hercules-ci.com/github/nix-community/cross-toolchains.nix/jobs/76

Comment (Mic92) [reply lib/systems/examples.nix:99]: I fixed this one.

File: lib/systems/examples.nix [hunk 2]
```diff
@@ -98,20 +98,14 @@ rec {
...
   mipsel-linux-gnu              = { config = "mipsel-linux-gnu";              } // platforms.gcc_mips32r2_o32;
-  mipsisa32r6-linux-gnu         = { config = "mipsisa32r6-linux-gnu";         } // platforms.gcc_mips32r6_o32;
-  mipsisa32r6el-linux-gnu       = { config = "mipsisa32r6el-linux-gnu";       } // platforms.gcc_mips32r6_o32;
 
   # require 64bit chip (for more registers, 64-bit floating point, 64-bit "long long") but use 32bit pointers
   mips64-linux-gnuabin32        = { config = "mips64-linux-gnuabin32";        } // platforms.gcc_mips64r2_n32;
   mips64el-linux-gnuabin32      = { config = "mips64el-linux-gnuabin32";      } // platforms.gcc_mips64r2_n32;
```
Comment (Mic92) [inline lib/systems/examples.nix:104]: Gcc does not know about this? https://hercules-ci.com/accounts/github/nix-community/derivations/%2Fnix%2Fstore%2F4ggih217yc0s4i7xs1wz3rdf1y71kj1c-mipsel-linux-gnu-stage-static-gcc-debug-10.3.0.drv/log?via-job=1df40d2a-d53d-4f4d-a872-9f43d9f87a85

Comment (Mic92) [reply lib/systems/examples.nix:104]: clang also fails:

nix-repl> :b pkgsCross.mips64-linux-gnuabin32.buildPackages.clang
error: build of '/nix/store/zjwxrsx8vjmhs1yqs8jy8973zcrkdf8f-compiler-rt-libc-mips64-linux-gnuabin32-11.1.0.drv' on 'ssh-ng://nix@ryan.r' failed: error: builder for '/nix/store/zjwxrsx8vjmhs1yqs8jy8973zcrkdf8f-compiler-rt-libc-mips64-linux-gnuabin32-11.1.0.drv' failed with exit code 2;
              last 10 log lines:
              > /build/compiler-rt-11.1.0.src/lib/asan/asan_interceptors_memintrinsics.h:53:10: warning: 'local_iovec.__sanitizer::__sanitizer_iovec::iov_base' may be used uninitialized in this function [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wmaybe-uninitialized-Wmaybe-uninitialized8;;]
              >    53 |     uptr __offset = (uptr)(offset);                                     \
              >       |          ^~~~~~~~
              > In file included from /build/compiler-rt-11.1.0.src/lib/asan/asan_interceptors.cpp:172:
              > /build/compiler-rt-11.1.0.src/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:3261:21: note: 'local_iovec.__sanitizer::__sanitizer_iovec::iov_base' was declared here
              >  3261 |   __sanitizer_iovec local_iovec;
              >       |                     ^~~~~~~~~~~
              > [ 80%] Built target RTAsan.mips64
              > [ 80%] Built target RTAsan_dynamic.mips64
              > make: *** [Makefile:136: all] Error 2
              For full logs, run 'nix log /nix/store/zjwxrsx8vjmhs1yqs8jy8973zcrkdf8f-compiler-rt-libc-mips64-linux-gnuabin32-11.1.0.drv'.
error: builder for '/nix/store/zjwxrsx8vjmhs1yqs8jy8973zcrkdf8f-compiler-rt-libc-mips64-linux-gnuabin32-11.1.0.drv' failed with exit code 1
error: 1 dependencies of derivation '/nix/store/ddvwbzlfnm0h06829gy3gzklqd4mg3kc-mips64-linux-gnuabin32-clang-wrapper-11.1.0.drv' failed to build

Comment (Mic92) [reply lib/systems/examples.nix:104]: fixed.

**Comment by a-m-joseph** [comment]
This is my first time using github to do a code review; if I did something stupid I apologize.

All of my review comments have an orange "Pending" on them.  I'm not sure what that means.  Am I supposed to do something now?

**Comment by a-m-joseph** [comment]
Hi, I can't figure out how to suggest specific changes.

Please consider the changes found in the topmost commit of this branch:

https://github.com/a-m-joseph/nixpkgs/tree/mic92-fix-platforms

A lot of the messiness here is because nixpkgs treats gnu 5-tuples as 4-tuples, and mixes up the libc with the abi.  So we have a lot of code duplication where adding an abi means you have to add it multiple times, once for each libc.

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by a-m-joseph** [comment]
> https://hercules-ci.com/github/nix-community/cross-toolchains.nix/jobs/75

Can you provide a `nix -E` expression that I can use to see if I've fixed the problem you're experiencing?

**Review by ghost — CHANGES_REQUESTED** [review]
Still can't figure out how to tell github that I suggest [these](https://github.com/a-m-joseph/nixpkgs/commit/f0af3de7b03d4482e87320854d22ee3e4d59111f) changes.

```
$ git push git@github:NixOS/nixpkgs pull/165020/head
ERROR: Permission to NixOS/nixpkgs.git denied to a-m-joseph.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

File: lib/systems/examples.nix [hunk 3]
```diff
@@ -98,20 +98,14 @@ rec {
   # can execute on 32bit chip
   mips-linux-gnu                = { config = "mips-linux-gnu";                } // platforms.gcc_mips32r2_o32;
   mipsel-linux-gnu              = { config = "mipsel-linux-gnu";              } // platforms.gcc_mips32r2_o32;
-  mipsisa32r6-linux-gnu         = { config = "mipsisa32r6-linux-gnu";         } // platforms.gcc_mips32r6_o32;
-  mipsisa32r6el-linux-gnu       = { config = "mipsisa32r6el-linux-gnu";       } // platforms.gcc_mips32r6_o32;
```
[the remark this answers is not in the document]
Comment (Mic92) [reply lib/systems/examples.nix:102]: This target does not evaluate at all. I am strongly against keeping broken targets unless you can show me how this target can be used.

File: lib/systems/platforms.nix [hunk 4]
```diff
@@ -490,8 +490,8 @@ rec {
   };
 
   # can execute on 32bit chip
-  gcc_mips32r2_o32 = { gcc = { arch = "mips32r2"; abi = "o32"; }; };
-  gcc_mips32r6_o32 = { gcc = { arch = "mips32r6"; abi = "o32"; }; };
+  gcc_mips32r2_o32 = { gcc = { arch = "mips32r2"; }; };
+  gcc_mips32r6_o32 = { gcc = { arch = "mips32r6"; }; };
```
[the remark this answers is not in the document]
Comment (Mic92) [reply lib/systems/platforms.nix:494]: The commit messages points to an gentoo issue, where they choose a similar ABI: https://bugs.gentoo.org/515694#c2
With the current abi target gcc does not compile.

File: lib/systems/doubles.nix [hunk 5]
```diff
@@ -28,6 +28,7 @@ let
     "aarch64-linux" "armv5tel-linux" "armv6l-linux" "armv7a-linux"
     "armv7l-linux" "i686-linux" "m68k-linux" "mipsel-linux" "mips64el-linux"
     "powerpc64-linux" "powerpc64le-linux" "riscv32-linux"
+    "mips-linux" "mips64-linux"
```
[the remark this answers is not in the document]
Comment (Mic92) [reply lib/systems/doubles.nix:31]: They evaluate and build stuff, see https://hercules-ci.com/github/nix-community/cross-toolchains.nix/jobs/79

**Comment by Mic92** [comment]
> > https://hercules-ci.com/github/nix-community/cross-toolchains.nix/jobs/75
> 
> Can you provide a `nix-env -E` expression that I can use to see if I've fixed the problem you're experiencing?

The code is here: https://github.com/nix-community/cross-toolchains.nix/blob/main/flake.nix

File: lib/tests/systems.nix [hunk 6]
```diff
@@ -17,7 +17,7 @@ with lib.systems.doubles; lib.runTests {
 
   testarm = mseteq arm [ "armv5tel-linux" "armv6l-linux" "armv6l-netbsd" "armv6l-none" "armv7a-linux" "armv7a-netbsd" "armv7l-linux" "armv7l-netbsd" "arm-none" "armv7a-darwin" ];
   testi686 = mseteq i686 [ "i686-linux" "i686-freebsd" "i686-genode" "i686-netbsd" "i686-openbsd" "i686-cygwin" "i686-windows" "i686-none" "i686-darwin" ];
-  testmips = mseteq mips [ "mips64el-linux" "mipsel-linux" "mipsel-netbsd" ];
+  testmips = mseteq mips [ "mips-linux" "mips64-linux" "mips64el-linux" "mipsel-linux" "mipsel-netbsd" ];
```
[the remark this answers is not in the document]
Comment (Mic92) [reply lib/tests/systems.nix:20]: see above.

File: pkgs/development/compilers/go/2-dev.nix [hunk 7]
```diff
@@ -40,6 +40,8 @@ let
     "armv6l" = "arm";
     "armv7l" = "arm";
     "powerpc64le" = "ppc64le";
+    "mips64" = "mips64";
+    "mips64el" = "mips64le";
```
Comment (zowoq) [inline pkgs/development/compilers/go/2-dev.nix:44]: Please drop the changes to `2-dev.nix` from this PR. This list of aches is out of order/sync with the other `go_1_` files. 

This derivation needs several other updates to match the others, I'll do all of it in a separate PR.

File: lib/systems/doubles.nix [hunk 8]
```diff
@@ -91,6 +91,7 @@ in {
   gnu           = filterDoubles (matchAttrs { kernel = parse.kernels.linux; abi = parse.abis.gnu; })
                   ++ filterDoubles (matchAttrs { kernel = parse.kernels.linux; abi = parse.abis.gnueabi; })
                   ++ filterDoubles (matchAttrs { kernel = parse.kernels.linux; abi = parse.abis.gnueabihf; })
+                  ++ filterDoubles (matchAttrs { kernel = parse.kernels.linux; abi = parse.abis.gnuabio32; })
```
Comment (Mic92) [inline lib/systems/doubles.nix:94]: Adding this commit did not fix evaluation:
https://hercules-ci.com/github/nix-community/cross-toolchains.nix/jobs/80

Comment (Mic92) [reply lib/systems/doubles.nix:94]:
```
nix-repl .
nix-repl> :b pkgsCross.mipsisa32r6-linux-gnu.stdenv
error: Unknown CPU type: mipsisa32r6
```

File: pkgs/development/compilers/go/2-dev.nix (hunk 7 above)
Comment (Mic92) [reply pkgs/development/compilers/go/2-dev.nix]: done.

File: lib/systems/examples.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (Mic92) [reply lib/systems/examples.nix]: If they are not meant to be used yet, there should be at least a comment indicating this.
Also the commit messages does not make any comments about them not being functional.

**Comment by Mic92** [comment]
No longer interested in investing time into MIPS.

File: lib/systems/platforms.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (a-m-joseph) [reply lib/systems/platforms.nix]: See #170736

**Comment by a-m-joseph** [comment]
See https://github.com/NixOS/nix/pull/6456#issuecomment-1111822528

Cross-builds to mips32 are now working, and the nix interpreter seccomp filters have been updated to allow mips32 binaries to execute in a sandbox on a mips64 host.

**Pull request closed without merging**